plugins/keyboard-navigation/src/navigation.js

```diff
@@ -380,4 +380,505 @@ export class Navigation {
+    }
+
+    if (!toolbox.getSelectedItem()) {
+      // Find the first item that is selectable.
+      const toolboxItems = toolbox.getToolboxItems();
+      for (let i = 0, toolboxItem; (toolboxItem = toolboxItems[i]); i++) {
+        if (toolboxItem.isSelectable()) {
+          toolbox.selectItemByPosition(i);
+          break;
+        }
+      }
+    }
+  }
+
+  /**
+   * Sets the navigation state to flyout and moves the cursor to the first
+   * block in the flyout.
+   * @param {!Blockly.WorkspaceSvg} workspace The workspace the flyout is on.
+   * @package
+   */
+  focusFlyout(workspace) {
+    const flyout = workspace.getFlyout();
+
+    this.setState(workspace, Constants.STATE.FLYOUT);
+
+    if (!this.getMarker(workspace).getCurNode()) {
+      this.markAtCursor(workspace);
+    }
+
+    if (flyout && flyout.getWorkspace()) {
+      const topBlocks = flyout.getWorkspace().getTopBlocks(true);
+      if (topBlocks.length > 0) {
+        const astNode = Blockly.ASTNode.createStackNode(topBlocks[0]);
+        this.getFlyoutCursor(workspace).setCurNode(astNode);
+      }
+    }
+  }
+
+  /**
+   * Sets the navigation state to workspace and moves the cursor to either the
+   * top block on a workspace or to the workspace.
+   * @param {!Blockly.WorkspaceSvg} workspace The workspace to focus on.
+   * @package
+   */
+  focusWorkspace(workspace) {
+    Blockly.hideChaff();
+    const reset = !!workspace.getToolbox();
+
+    this.resetFlyout(workspace, reset);
+    this.setState(workspace, Constants.STATE.WORKSPACE);
+    this.setCursorOnWorkspaceFocus(workspace);
+  }
+
+  /**
+   * Gets the cursor on the flyout's workspace.
+   * @param {!Blockly.WorkspaceSvg} workspace The main workspace the flyout is
+   *     on.
+   * @return {?Blockly.FlyoutCursor} The flyout's cursor or null if no flyout
+   *     exists.
+   * @protected
+   */
+  getFlyoutCursor(workspace) {
+    const flyout = workspace.getFlyout();
+    const cursor = flyout ? flyout.getWorkspace().getCursor() : null;
+
+    return /** @type {Blockly.FlyoutCursor} */ (cursor);
+  }
+
+  /**
+   * Hides the flyout cursor and optionally hides the flyout.
+   * @param {!Blockly.WorkspaceSvg} workspace The workspace.
+   * @param {boolean} shouldHide True if the flyout should be hidden.
+   * @protected
+   */
+  resetFlyout(workspace, shouldHide) {
+    if (this.getFlyoutCursor(workspace)) {
+      this.getFlyoutCursor(workspace).hide();
+      if (shouldHide) {
+        workspace.getFlyout().hide();
+      }
+    }
+  }
+
+  /**
+   * Tries to connect the given marker and cursor node.
+   * @param {!Blockly.WorkspaceSvg} workspace The main workspace.
+   * @param {!Blockly.ASTNode} markerNode The node to try to connect to.
+   * @param {!Blockly.ASTNode} cursorNode The node to connect to the markerNode.
+   * @return {boolean} True if the key was handled; false if something went
+   *     wrong.
+   * @protected
+   */
+  modify(workspace, markerNode, cursorNode) {
```

Comment:
I feel this name could be more descriptive (I'm not actually sure what's being modified here) and moving it might be a good time to change it, but also fine if you don't want to change things.

Reply:
Changed to tryToConnectMarkerAndCursor. Let me know what you think.

Reply:
Makes sense :) thanks!